recurring_contract/models/recurring_contract.py

```diff
@@ -182,39 +185,54 @@ def unlink(self):
     #                             PUBLIC METHODS                             #
     ##########################################################################
     @api.multi
-    def clean_invoices(self, since_date=None, to_date=None, keep_lines=None):
+    def clean_invoices(self, since_date=None, to_date=None, keep_lines=None, **kwargs):
         """ By default, launch asynchronous job to perform the task.
             Context value async_mode set to False can force to perform
             the task immediately.
         """
         if self.env.context.get('async_mode', True):
-            self.with_delay()._clean_invoices(since_date, to_date, keep_lines)
+            self.with_delay()._clean_invoices(since_date, to_date, keep_lines, **kwargs)
         else:
-            self._clean_invoices(since_date, to_date, keep_lines)
+            self._clean_invoices(since_date, to_date, keep_lines, **kwargs)
 
     def rewind_next_invoice_date(self):
         """ Rewinds the next invoice date of contract after the last
         generated invoice. No open invoices exist after that date. """
         gen_states = self.env['recurring.contract.group']._get_gen_states()
+
+        res = self.env["account.invoice"]
+
         for contract in self:
             if contract.state in gen_states:
-                last_invoice_date = max([
-                    line.invoice_id.date_invoice for
-                    line in contract.invoice_line_ids
-                    if line.state in ('open', 'paid')] or [False])
+                last_invoice_date, is_paid = min(
+                    [(line.invoice_id.date_invoice, line.state == "paid") for
+                     line in contract.invoice_line_ids
+                     if line.state in ("open", "paid") and
+                     line.invoice_id.date_invoice >= date.today()] or [
+                        (False, False)])
+
+                if last_invoice_date and is_paid:
+                    last_invoice_date += contract.group_id.get_relative_delta()
```

Comment:
What if there are 3 future invoices in june, july and august? Your code would take june as `last_invoice_date` and then use july if june is paid. But July may also be paid. That might be an issue.

Reply:
True. I didn't see this issue

Reply:
I'll add a testcase for this scenario and see what changes need to be made